core/src/main/java/com/spotify/metrics/core/DistributionImpl.java

```diff
@@ -0,0 +1,75 @@
+/*
+ * Copyright (c) 2016 Spotify AB.
+ *
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+package com.spotify.metrics.core;
+
+
+import com.google.common.annotations.VisibleForTesting;
+import com.spotify.metrics.core.codahale.metrics.ext.Distribution;
+import com.tdunning.math.stats.TDigest;
+
+import java.nio.ByteBuffer;
+import java.util.concurrent.atomic.AtomicReference;
+
+
+public class DistributionImpl implements Distribution {
+
+    private static final int COMPRESSION_DEFAULT_LEVEL = 100;
+    private final AtomicReference<TDigest> distRef;
+
+    protected DistributionImpl() {
+        this.distRef = new AtomicReference<>(TDigest.createDigest(COMPRESSION_DEFAULT_LEVEL));
+    }
+
+    @Override
+    public void record(double val) {
+        distRef.get().add(val);
```

Comment:
TDigest.add() does not appear to be threadsafe

Reply:
Yeah, I am looking into this. I checked couple of Dropwizard metrics there are threadsafe so I will make this thread safe.

Reply:
I think we could either
* add synchronized around it and hope it's enough.
* Use ThreadLocal<TDigest> and then merge all the thread local entries somehow inside getValueAndFlush
* Push to a queue and have another thread popping them off and adding to the TDigest object

Reply:
Please check the implementation I added yesterday.  I thought about adding threadLocal then change mind.  ThreadLocal may introduce memory leak if getValueAndFlush isn't called often.

Reply:
Yes, that should work now I think (but it's hard to know if there will be much contention, some multithreaded JMH benchmarks might be useful).

Reply:
Ok, sure will do.  Please take a look I need a +1